### PR TITLE
APEXMALHAR-2366 #resolve #comment Apply BloomFilter to Bucket, use in…

### PR DESCRIPTION
**Pull request by brightchen**
…ternal BloomFilter

### CODE REVIEW

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 1]
```diff
@@ -353,6 +362,41 @@ public Slice get(Slice key, long timeBucket, ReadSource readSource)
       }
     }
 
+
+    private int filtedCount = 0;
```
Comment (davidyan74) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java]: Check spelling

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 2]
```diff
@@ -353,6 +362,41 @@ public Slice get(Slice key, long timeBucket, ReadSource readSource)
...
+    private int unfiltedCount = 0;
```
Comment (davidyan74) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java]: Check spelling

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 3]
```diff
@@ -353,6 +362,41 @@ public Slice get(Slice key, long timeBucket, ReadSource readSource)
...
+     * @param mightContain true if collection might contain the value
+     */
+    private void verifyBloomFilter(boolean mightContain)
+    {
+      if (!mightContain) {
+        filtedCount++;
+      } else {
+        unfiltedCount++;
+      }
+      long now = System.currentTimeMillis();
```
Comment (davidyan74) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java]: Would this introduce problems with idempotency?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+
+import java.util.BitSet;
+
+import com.datatorrent.netlet.util.Slice;
+
+/**
+ * This class implemented BloomFilter algorithm, the key is Slice
+ *
+ */
+public class SliceBloomFilter
```
Comment (davidyan74) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java:29]: Is the code of this class copied from somewhere? If so, is there a reason why we don't use it as a library?

**Review by davidyan74 — CHANGES_REQUESTED** [review]

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java (hunk 4 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java:29]: @davidyan74 
Yes, I copied from org.apache.hadoop.util.bloom.BloomFilter and modified. The reason of copy/modify instead of using are
- The original BloomFilter can't be serialized.
- The original BloomFilter use temporary memory, this one use SerializationBuffer

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 3 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java]: @davidyan74 this change only impact performance but don't impact functional. So I think it does not break the idempotency.

**Comment by brightchen** [comment]
@davidyan74 applied comments, please review

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java (hunk 4 above)
Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java:29]: It would be good if you include that information in the javadoc of this class. Let me spend some time later to look at this more. Since this has only to do with performance and does not affect the functionality of the Windowed Operator, this can wait, correct?

Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java:29]: @brightchen is there a reason why we are not using Guava's Bloom Filter implementation?

**Comment by bhupeshchawda** [comment]
@brightchen Added some comments in the JIRA

**Comment by brightchen** [comment]
@bhupeshchawda commented in the jira

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 5]
```diff
@@ -232,6 +232,12 @@ public int hashCode()
 
     protected ConcurrentLinkedQueue<Long> windowsForFreeMemory = new ConcurrentLinkedQueue<>();
 
+    private static boolean disableBloomFilterByDefault = false;
```
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:235]: Do we need this constant? Can't we directly use `TRUE`/ `FALSE` instead?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 6]
```diff
@@ -232,6 +232,12 @@ public int hashCode()
...
+    private boolean disableBloomFilter = disableBloomFilterByDefault;
```
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:236]: Rename to `enableBloomFilter`? Also can we enable it by default? `enableBloomFilter = TRUE`

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 7]
```diff
@@ -232,6 +232,12 @@ public int hashCode()
...
+    private static int bloomFilterDefaultBitSize = 1000000;
```
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:237]: If this is supposed to be a constant, can we make it something like: `DEFAULT_BLOOM_FILTER_SIZE`?
Also make all constants final?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 8]
```diff
@@ -353,6 +362,41 @@ public Slice get(Slice key, long timeBucket, ReadSource readSource)
       }
     }
 
+
+    private int filteredCount = 0;
+    private int unfilteredCount = 0;
+    private long verifyBloomFilterBeginTime = 0;
+    private long verifyBloomFilterInterval = 5000;
```
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java]: Can we move these declarations to the top?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 9]
```diff
@@ -353,6 +362,41 @@ public Slice get(Slice key, long timeBucket, ReadSource readSource)
...
+    /**
+     * Test the result of using bloom filter and remove it if it not helpful for improving the performance.
+     *
+     * @param mightContain true if collection might contain the value
+     */
+    private void verifyBloomFilter(boolean mightContain)
```
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:374]: I think this might not be the right way to see if the bloom filter is performing well. Since the number of elements added to the bloom filter will affect the false positive prob., can you try to use `bloomFilter.expectedFalsePositiveProbability()` to see when we can stop using the bloom filter?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 10]
```diff
@@ -411,6 +465,7 @@ public void put(Slice key, long timeBucket, Slice value)
     {
       // This call is lightweight
       releaseMemory();
+
```
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:466]: unnecessary change

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 11]
```diff
@@ -361,6 +405,16 @@ public Slice get(Slice key, long timeBucket, ReadSource readSource)
      */
     private BucketedValue getValueFromTimeBucketReader(Slice key, long timeBucket)
     {
+      if (bloomFilter != null) {
```
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java]: I think we should use the bloom filter at the first `get` call.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 12]
```diff
@@ -454,6 +509,14 @@ public long freeMemory(long windowId) throws IOException
         entryIter.remove();
 
         for (Map.Entry<Slice, BucketedValue> entry : windowData.entrySet()) {
+          /**
+           * The data still in memory and reachable before the memory released
+           * So put key into bloom filter here
+           */
+          if (bloomFilter != null) {
```
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:514]: Adding an element to the bloom filter is best done when the element is fetched. This should also be in the first `get` call.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 13]
```diff
@@ -506,6 +569,7 @@ protected long releaseMemory()
     public Map<Slice, BucketedValue> checkpoint(long windowId)
     {
       releaseMemory();
+
```
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:569]: unnecessary

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 14]
```diff
@@ -630,6 +694,49 @@ public WindowedBlockStream getValueStream()
...
+    public void setDisableBloomFilter(boolean disableBloomFilter)
+    {
+      this.disableBloomFilter = disableBloomFilter;
+      this.unloadBloomFilter();
+    }
+
+
     private static final Logger LOG = LoggerFactory.getLogger(DefaultBucket.class);
+
```
Comment (bhupeshchawda) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java]: avoid unnecessary changes

File: library/src/test/java/org/apache/apex/malhar/lib/state/managed/DefaultBucketTest.java [hunk 15]
```diff
@@ -52,6 +53,9 @@
     @Override
     protected void starting(Description description)
     {
+      //lots of test case get around the normal workflow and directly write to file. So should disable bloom filter
+      DefaultBucket.setDisableBloomFilterByDefault(true);
```
Comment (bhupeshchawda) [inline library/src/test/java/org/apache/apex/malhar/lib/state/managed/DefaultBucketTest.java:57]: Don't understand the reason why the bloom filter is disabled here?

File: library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedTimeStateImplTest.java [hunk 16]
```diff
@@ -81,6 +83,7 @@ public void testAsyncGetFromReaders() throws IOException, ExecutionException, In
     Slice zero = ManagedStateTestUtils.getSliceFor("0");
     long time = System.currentTimeMillis();
 
+    DefaultBucket.setDisableBloomFilterByDefault(true);
```
Comment (bhupeshchawda) [inline library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedTimeStateImplTest.java:88]: Same here, why disable?

File: library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedTimeUnifiedStateImplTest.java [hunk 17]
```diff
@@ -116,6 +118,8 @@ public void testAsyncGet() throws ExecutionException, InterruptedException
   @Test
   public void testSyncGetFromFiles() throws IOException, ExecutionException, InterruptedException
   {
+    DefaultBucket.setDisableBloomFilterByDefault(true);
```
Comment (bhupeshchawda) [inline library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedTimeUnifiedStateImplTest.java:121]: Same comment.

**Review by bhupeshchawda — CHANGES_REQUESTED** [review]

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 5 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:235]: My intension is giving the client code a choice to use or disable the bloom filter for all bucket. As we don't have a central place to manage the instances. It's hard to get the instance for app developer. I'll change it into public

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 6 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:236]: I treat enable bloom filter as the default. that's why use disableBloomFilter for special case

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 7 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:237]: I give a chance for the application developer to change the default bit size in class level instead of instance level.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 8 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java]: The reason I put them here is all of them in fact only used by method verifyBloomFilter()

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 9 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:374]: I'll look into it.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 11 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java]: First get is get from memory, which don't take too much time. And if we take check for the first get, which mean we need to put the flag at the first put(), which probably decrease the accurate very quickly

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 12 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:514]: The flag will be put when read data from files.
Let assume everything is in memory, do we still need bloom filter? probably not as it's simple a map. So I think filter only applies to decrease the chance of unnecessary reading from files

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 5 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:235]: I already provides setter/getter, so private is ok

File: library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedTimeStateImplTest.java (hunk 16 above)
Comment (brightchen) [reply library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedTimeStateImplTest.java:88]: the bloom filter works only go through put() and get() method of bucket. But this unit test case write data direct data from file instead of go through put(), so the bloom filter didn't keep the flags and get() return null due to filtered by Bloom Filter.

File: library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedTimeUnifiedStateImplTest.java (hunk 17 above)
Comment (brightchen) [reply library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedTimeUnifiedStateImplTest.java:121]: same as previous one.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 13 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:569]: just separate the logic section

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 10 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:466]: just separate the logic section

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 5 above)
Comment (bhupeshchawda) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:235]: If a centralized place is needed, wouldn't AbstractManagedStateImpl be a better place to enable / disable managed state?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 6 above)
Comment (bhupeshchawda) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:236]: I still think the boolean should be named `enableBloomFilter`. You can set it to true and move it to the AbstractManagedStateImpl instead of Bucket.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 7 above)
Comment (bhupeshchawda) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:237]: I meant the constant value set to 1000000 should be a static final field. In fact all defaults must be public static final constants and set in the AbstractManagedStateImpl, instead of Bucket.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 11 above)
Comment (bhupeshchawda) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java]: How about the `getFromReaders` call then? That can be a place for using the bloom filter. Also once the value is fetched, it should immediately be inserted into the bloom filter. I don't understand how that would decrease the accuracy of the filter. Can you clarify the second part of your comment?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 12 above)
Comment (bhupeshchawda) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:514]: I agree. Can we use `getFromReaders` in that case to insert into bloom filter? I don't think `freeMemory` is the right place to put values in the bloom filter.

File: library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedTimeStateImplTest.java (hunk 16 above)
Comment (bhupeshchawda) [reply library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedTimeStateImplTest.java:88]: I understand. In that case, I think we should have a unit test which actually executes the code changes done for supporting bloom filter. Probably a test which inserts a fixed number of keys into the store and uses bloom filter to see how much can be fetched without accessing the disk.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java (hunk 4 above)
Comment (bhupeshchawda) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java:29]: @davidyan74 are you okay with the code for bloom filter that @brightchen has taken from  org.apache.hadoop.util.bloom.BloomFilter? We'll definitely add this part into the javadoc of the class. Is there anything else that you wanted to look into?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 5 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:235]: What I meant about instance management is something like Spring which can easily configure the property of instances. As we don't have this mechanism, so the client code very difficult to get the instance(for this case is Bucket) to configure the property. That's why provide a static default value, client code can't easily change the default value. The unit test ManagedTimeStateImplTest.testAsyncGetFromReaders() is one example

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 6 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:236]: Different developer has different preference. It's not really impact anything like performance or complexity. 
My logic is:
if(special case)  // in this case is disabled bloom filter as treat bloom filter as default behavior
{
  do something
}

instead of 
if(not special case) 
{
  do something
}

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 7 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:237]: But the bloomFilterDefaultBitSize is changeable.
And why put it into AbstractManagedStateImpl? this variable is totally depends on the implementation. Some implementation probably don't have conception at all

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 11 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java]: Why the value should be inserted into the bloom filter after fetched? Do you try to avoid update with same value? Could you explain more here?

What I usage is different here. What I try to avoid to is unnecessary get as there are much more getting than setting. 

The reason why decrease accurate is due to there are much more put() in all than put() to the file. and each put add flags to the bloom filter,

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 12 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:514]: I guess this is due to misunderstanding the usage of bloom filter.
freeMemory() is only place which we know data going to write to files

File: library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedTimeStateImplTest.java (hunk 16 above)
Comment (brightchen) [reply library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedTimeStateImplTest.java:88]: Agree.
I'll add unit test case

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 9 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:374]: expectedFalsePositiveProbability() depends on expectedNumberOfFilterElements, which we in fact have no idea at all. At it is a fix value instead of dynamic value

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 11 above)
Comment (bhupeshchawda) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java]: Agreed. Makes sense.
Can we move the usage of bloom filter to the `getFromReaders()` call? That is the first place where we know that files might be accessed.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 9 above)
Comment (bhupeshchawda) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:374]: Sorry, I meant using `getFalsePositiveProbability()`. We can keep track of the number of keys inserted into the bloom filter, right?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 12 above)
Comment (bhupeshchawda) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:514]: agreed

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 7 above)
Comment (bhupeshchawda) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:237]: I think the properties which allow to configure a bloom filter should not be per bucket. I am talking about variables like: `disableBloomFilter` and `bloomFilterBitSize`. I think these should be kept configurable. Would the user ever want to configure these differently for different buckets? How would a user of managed state be able to configure the bloom filter size?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 5 above)
Comment (bhupeshchawda) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:235]: Similar comment as below.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java (hunk 4 above)
Comment (bhupeshchawda) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java:29]: @brightchen Can you please also add the source of the code in javadoc?

File: library/src/test/java/org/apache/apex/malhar/lib/state/managed/DefaultBucketTest.java [hunk 18]
```diff
@@ -227,4 +233,33 @@ public void testFreeMemory() throws IOException
 
     testMeta.defaultBucket.teardown();
   }
+
+  @Test
+  public void testBloomFilter() throws IOException
```
Comment (bhupeshchawda) [inline library/src/test/java/org/apache/apex/malhar/lib/state/managed/DefaultBucketTest.java:238]: This test would succeed whether or not bloom filter is enabled. Can we do some sort of validation whether the bloom filter in the bucket is consistent with the results?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 19]
```diff
@@ -247,6 +253,9 @@ public DefaultBucket(long bucketId)
     public void setup(@NotNull ManagedStateContext managedStateContext)
     {
       this.managedStateContext = Preconditions.checkNotNull(managedStateContext, "managed state context");
+      if (!disableBloomFilter && bloomFilter == null) {
+        bloomFilter = new SliceBloomFilter(bloomFilterBitSize, 0.99);
+      }
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:258]: I think "bloomFilter" has to move to time bucket because bloomFilter under DefaultBucket will work only if the managedState is of type ManagedTimeUnifiedStateImpl. Here, DefaultBucket behaves as time Bucket.
If the managedState is of type ManagedStateImpl/ManagedTimeStateImpl then DefaultBucket behaves as key bucket.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 7 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:237]: In most of case, disableBloomFilter and bloomFilterBitSize should not be per bucket. It's kind depend the type of bucket. If this is the case, the static attributes can be used.
What I concern about put them into state manager is the it totally depends on Bucket implementation and not related to state manager at all. Let's think about we have another implementation, such as MemoryBucket, these attributes are non-sense to this implementation at all.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 9 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:374]: I did very simple test, getFalsePositiveProbability() in fact is not very light. it cost 142ms for 1M operation.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 19 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:258]: Yes, for ManagedStateImpl, bloom filter would be useless for a long run(But I still useful for certain period). ManagedTimeStateImpl should be useful for most of time. 
That's why I add the simple mechanism to check the efficiency of bloomfilter to dynamic unload it if it is useless.

File: library/src/test/java/org/apache/apex/malhar/lib/state/managed/DefaultBucketTest.java (hunk 18 above)
Comment (brightchen) [reply library/src/test/java/org/apache/apex/malhar/lib/state/managed/DefaultBucketTest.java:238]: Yes, bloom filter should only helpful to increase the performance which should be tested in Benchmark application as it need run for a long time . I think what we should test in unit test is apply bloom filter will not break the current result.

**Comment by brightchen** [comment]
@bhupeshchawda @chaithu14 I think I applied all comments or comment back. Please let me know if any leftover or new comments

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 20]
```diff
@@ -353,6 +362,32 @@ public Slice get(Slice key, long timeBucket, ReadSource readSource)
...
+    private void verifyBloomFilter(boolean mightContain)
+    {
+      if (!mightContain) {
+        filteredCount++;
+      } else {
+        unfilteredCount++;
+      }
+
+      if (unfilteredCount > filteredCount * 4) {
+        unloadBloomFilter();
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java]: I think this looks like not a feasible way of disabling bloomFilter. Consider the case where the first search element(which would be on disk) found in state then the if condition becomes true and then the bloomFilter is disabled.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java [hunk 21]
```diff
@@ -0,0 +1,459 @@
...
+{
+  private BitSet bitset;
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java:31]: BitSet is not Kryo serializable. Could you please bind it under JavaSerializer?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java [hunk 22]
```diff
@@ -0,0 +1,459 @@
...
+      SetAttributes(Math.ceil(-(Math.log(falsePositiveProbability) / Math.log(2))) / Math.log(2), // c = k / ln(2)
+          expectedNumberOfElements, (int)Math.ceil(-(Math.log(falsePositiveProbability) / Math.log(2))));
+    }
+  }
+
+  /**
+   * Generate integer array based on the hash function till the number of
+   * hashes.
+   *
+   * @param data
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java]: Mismatching with the argument. Could you please correct it.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java [hunk 23]
```diff
@@ -0,0 +1,459 @@
...
+  public void clear()
+  {
+    bitset.clear();
+    numberOfAddedElements = 0;
+  }
+
+  /**
+   * Adds an array of bytes to the Bloom filter.
+   *
+   * @param bytes
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java]: Same here, update the comment.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java [hunk 24]
```diff
@@ -0,0 +1,459 @@
...
+    }
+    numberOfAddedElements++;
+  }
+
+  /**
+   * Returns true if the array of bytes could have been inserted into the Bloom
+   * filter. Use getFalsePositiveProbability() to calculate the probability of
+   * this being correct.
+   *
+   * @param bytes
```
Comment (chaithu14) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java]: Same here

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 7 above)
Comment (bhupeshchawda) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:237]: Then we should probably remove the setter methods for disableBloomFilter and bloomFilterBitSize, right?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 9 above)
Comment (bhupeshchawda) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java:374]: I think that should be okay. We don't have to check this every time. We can do it once every n elements inserted, right?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java (hunk 20 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java]: agree. This logic is based on statistics, so the base number should be relative big. I'll add logic to check base number.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java (hunk 21 above)
Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java:31]: I believe I saw an implementation which was annotated by JavaSerializer and I then removed it. I can't recalled the reason now. But I check the BitSet class, it seems serializable by kryo. Could you please check the version of BitSet? I'll add a unit test case for building a simple application to test this. The unit test application should throw exception if any non-transient object is not serializable.

Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java:31]: I added unit test and no exception thrown. So it should be serializable

**Comment by brightchen** [comment]
@bhupeshchawda @chaithu14 Could you please merge if no further comments

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java (hunk 21 above)
Comment (chaithu14) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java:31]: :+1:

**Comment by brightchen** [comment]
@PramodSSImmaneni @vrozov I talked with @bhupeshchawda He want one of you take look at this PR

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java (hunk 4 above)
Comment (pramodin) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java:29]: I will add the reference

**Comment by pramodin** [comment]
It looks like this is close to merge and needs a couple of small things. Since @brightchen is not looking at this actively, I will make those changes on top of a merge commit containing his commit and create a new PR for same JIRA. This way the final merge will also close this PR.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java (hunk 4 above)
Comment (pramodin) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/SliceBloomFilter.java:29]: I notice that the original hadoop bloom filter code includes a second license from European Commission on top of the regular apache license. We may have to do the same. Will discuss this on the dev list.
